core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -949,8 +949,11 @@ class ReplicaManager(val config: KafkaConfig,
     else
       FetchHighWatermark
 
-    // Restrict fetching to leader if request is from follower or from a client with older version (no ClientMetadata)
-    val fetchOnlyFromLeader = isFromFollower || (isFromConsumer && clientMetadata.isEmpty)
+    // Restrict fetching to leader if request is from follower or from an ordinary consumer
+    // with an older version (which is implied by no ClientMetadata)
```

Comment:
Seems like this comment is a bit redundant. The main useful info is that no ClientMetadata means older version, could we have a method name that made that clear and then the comment would not be needed?

Reply:
The second question is: do we want to allow this debugging consumer id behavior at all? Is it useful in the new world?

Reply:
Yeah, I was debating the latter. It's probably unlikely that this is used outside of the ReplicaVerificationTest. As far as I know, no one has filed an issue about it, so maybe we could drop this support. I decided to err on the side of compatibility, but I could probably be convinced to change my mind.

Reply:
Can we remove `DebuggingConsumerId` altogether or is there any functionality that it provides in today's world? If it doesn't provide any useful functionality now, I would just remove it and it's one less concept to support and think about. As you said, I don't think anyone actually uses this outside of the `ReplicaVerificationTool`.

Reply:
If there are any use cases which rely on it, users can upgrade their clients to regain the ability. Since no one has evidently noticed that this functionality was lost, I think it's probably fine to remove it.

Reply:
Sounds good.